### DISCUSSION TITLE
Accessing geometries for fly-to functionality

### DISCUSSION THREAD

**Discussion by TylerMatteo** [NYCPlanning/ae-cp-map#43] [General]
In working on #16, I had trouble implementing this requirement:

> When user selects a projects, map is panned so that the selected project is centered

I have some thoughts on next steps, but the complexity seemed significant enough to break out of the original Issue and capture those thoughts in a discussion.

## Context
The requirements ask that the map should pan and zoom to center the selected the project. All parts of the project should be in the viewport and this functionality should be aware of route paths we're using to track the managing code and ID of the selected project. By that logic, if a user clicks a few projects in succession then clicks the back button in their browser, we would expect the map `viewstate` to move as they cycle backwards through the projects they selected.

We implemented the a pan-on-selection pattern in [ae-zoning-map-poc ](https://github.com/NYCPlanning/ae-zoning-map-poc/blob/e6d458ca8fd43d74166ae122a80de1150651e51c/src/App.tsx#L65) but there are a couple things to note about that implementation:
* That app doesn't have the added complexity of routing
* We made an API for returning individual tax lots as geojson features. This made it easier to update the viewstate based on the shape of the tax lot compared to this app where (so far) we only have access to the capital project geometries through the underlying vector tiles
* It uses the centroid of the tax lot to update viewstate with a hard coded `zoom`. This was fine for the proof of concept but for capital projects, we will need a more robust approach that gives us a proper bounding box of the project.

## What I tried
I had hoped we could accomplish this functionality using the geometries from the tiles and the various APIs Deck gives us but I don't think there's a solution that works consistently and is performant. If there is, it's going to introduce some seriously janky code. This issue really comes down to the nature of tiles. By their nature, they slice up the geoms of projects based on which tiles a project partially or fully lands in.

I made a [throwaway branch](https://github.com/NYCPlanning/ae-cp-map/blob/fly-to-project-poc/app/components/atlas.client.tsx) from my feature branch for #16 that shows the closest I could get to having this working. Don't look at this code too long, for eventually _it will stare back_. This approach runs MVTLayer's [getRenderedFeatures](https://deck.gl/docs/api-reference/geo-layers/mvt-layer#getrenderedfeatures) method once when the map first loads at the citywide viewport and puts that into state. That method is an expensive operation in the best of use cases, but this approach is especially bad because we're chucking thousands of project geoms into memory. What's worse, this approach seems to be flaky anyway, presumably because Deck is putting an upper bound on how many features it actually lets that method return and features it does return are still sliced up based on the tiling. Once I had that list of features, it's just a matter of grabbing the selected project feature based on the path params, getting its bounding box, and updating `viewstate` accordingly. [The documentation for fitBounds](https://deck.gl/docs/api-reference/core/web-mercator-viewport#fitbounds) was helpful for converting a bounding box to lat/long/zoom values.

## Solutions
Here are two approaches I see to solving this issue. Both require changes to ae-zoning-api

### 1. Add extend info to the existing capital project endpoint
We could add a field to the data returned by the existing `/capital-projects/:managing-code/:capital-project-id` endpoint containing the bounding box for the project expressed as `[[number, number],[number, number]]`. We could use PostGIS' [ST_Extent](https://postgis.net/docs/ST_Extent.html) function to do the calculation in the database. We could then use this to update `viewstate`. Once the view state is updated to fit the entire project, the tiles will handle actually rendering it. This approach is relatively easy to implement but I have complicated feelings about it - mostly that it feels like we'd be adding a new field to an API for a pretty specific use case, which is a slippery slope if we're trying to build client-agnostic APIs. Also, having lat/lon arrays return in otherwise non-geospatial objects feels icky. We have a standards for sending geospatial data over the network: GeoJSON. 

### 2. Add an endpoint to return projects as geojson features
Similar to the approach we took in the PoC app, we could have another endpoint that return a single project as a geojson feature. The geometries for projects can be much more complex than tax lots, so we'd have to keep an eye on how much geojson we're sending over the wire. Once we have geojson of the full project on the front end, there are two ways we could use it:
1. Simply use it to calculate a bounding box and update viewstate accordingly
2. Add a new `GeojsonLayer` that takes in the geojson feature of the selected project. This might seem like an odd approach since we already have the projects on the map from the tiles but I think it does have some benefits. For one, because that new layer will only be as big as the individual feature, we can use it's bounds to update viewstate instead of calculating a bbox with something like turf.js (that's what the example for [fitBounds](https://deck.gl/docs/api-reference/core/web-mercator-viewport#fitbounds) does where you see `layer.getBounds()`). Another potential upside is that we could use this layer to render highlight the selected project instead of doing it `getFillColor` of the projects tile layer. The current approach run `getFillColor` for all rendered projects every time a new project is selected. Rendering the single selected project "on top" of the tiles might actually be a snappier experience for users.

## Next steps
I'm thinking we move forward with option two. If we do, the next step would be to create the new project geojson endpoint, including the corresponding updates to the openAPI yaml. The `properties` of the geojson feature would reuse the existing `CapitalProjectBudgeted` schema. The geometry type would be dynamic based on the geom type for the given project.

**Comment by dhochbaum-dcp** [comment]
- Adding map coordinates to center and zoom the map might be a specific use case, but including the location of a project feels like a pretty generic use case, so I'm not against option 1.  
- Are we trying to build client-agnostic APIs?  Why?
- I'm having some difficulty conceptualizing what the difference would be in terms of number of API requests.  Is what you're describing in option 2 going to lead to network requests as we move the mouse around the map and hover over different projects?

**Reply by TylerMatteo** [reply]
> Are we trying to build client-agnostic APIs? Why?

Yes, at least as a general best practice. Decoupling APIs from specific client use cases can help keep them more reusable and fend off bloating them unnecessarily. Granted, this wouldn't be the most egregious case of adding a client-specific use case, but I do think it's leaning that way.

> I'm having some difficulty conceptualizing what the difference would be in terms of number of API requests. Is what you're describing in option 2 going to lead to network requests as we move the mouse around the map and hover over different projects?

Not quite. The code would call the new endpoint when a project is selected, either by clicking on the map or, when it's implemented, selecting it from a list of projects. Therefore, I wouldn't expect it to trigger any more network calls than the current implementation. The difference is the size of the response would be bigger because it includes the geometry for the project, instead of just its properties.

**Comment by TangoYankee** [comment]
I agree with option two for the reasons stated above.

However, when we consider the implementation of the capital projects geojson endpoint, I would like us to consider a deviation from how we implemented the tax lots geojson endpoint. The current implementation has a "geojson" path suffix ie) `taxlots/{bbl}/geojson`. For capital projects, I suggest using an Accept [Request header](https://developer.mozilla.org/en-US/docs/Glossary/Request_header) of `application/geo+json` and dropping the path suffix.

Regardless of whether we want to drop the `geojson` path suffix, we should be using `application/geo+json` as the Content Type for any endpoint that returns geojson. Geojson is structured different from json. Standard json stores the non-spatial data at the top-level of the object. Geojson stores the non-spatial data in a nested "properties" field.

If we're going to differentiate the responses based on the Accept/Content-Type headers, then we might as well rely on it as the sole differentiator. This will make the paths more concise

**Reply by TylerMatteo** [reply]
Does OpenAPI let you specify different response schemas based on the `Accepts` header? I recall going down the rabbit hole myself a while back and I didn't turn up anything. I don't love the `/geojson` approach either but any alternatives need to play nice with OAS and our toolchain. Regardless, I fully support adding those headers to any future `/geojson` endpoints.

**Reply by TangoYankee** [reply]
Yes, [comment linking to the relevant documentation](https://github.com/OAI/OpenAPI-Specification/issues/146#issuecomment-358665726) and also [documentation for header parameters](https://swagger.io/docs/specification/describing-parameters/#header-parameters)

A [Quick example for zoning-api](https://github.com/NYCPlanning/ae-zoning-api/compare/main...43/header-exp) This example uses a pre-existing "MultiPolygon" schema for a quick demonstration. The production version would define a full "CapitalProjectBudgetedSpatial" schema to return. 

On the documentation page, you'll see the "Response Schema" field turns into a dropdown to select the appropriate content type.

![response_schema_content_dropdown_geo](https://github.com/NYCPlanning/ae-cp-map/assets/11068062/c3c5cc13-74c0-464e-aa8b-7af7dc57a8f4)

**Reply by TylerMatteo** [reply]
I ran Kubb from that branch and, unfortunately, it doesn't look like it supports this 😞 it was just using the multipolygon schema for the response, presumably because it takes the last schema it sees in the YAML. We'll stick with the `/geojson` pattern until we come up with a more idiomatic solution that plays nice with OAS and our toolchain.

**Comment by pratishta** [comment]
I like Option 2 with option 2 (using a new GeojsonLayer). Would this be strictly for the project selections or can this layer also be used to highlight a selected district?

**Reply by TylerMatteo** [reply]
Good call out, I was thinking this might have implications for the district filter functionality. I was going to suggest we handle filtering on the map using Deck's [MaskExtension](https://deck.gl/docs/api-reference/extensions/mask-extension) but that might be tricky with the districts only existing as tiled data. Having a separate layer to render the full district as geojson would make it much easier. If we have that, we may as well use it to show the outline of the selected district.

**Comment by TylerMatteo** [comment]
Ok, I created issues to document and implement the new endpoint and then use it to add "fly to" functionality on the frontend:
* https://github.com/NYCPlanning/ae-zoning-api/issues/341
* https://github.com/NYCPlanning/ae-zoning-api/issues/342
* https://github.com/NYCPlanning/ae-cp-map/issues/49
